File: SocialNetwork/src/main/java/com/mini/socialnetwork/service/CommentService.java
package com.mini.socialnetwork.service;

import java.io.IOException;
import java.time.Instant;
import java.util.Map;
import java.util.UUID;

import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import org.springframework.web.multipart.MultipartFile;

import com.cloudinary.Cloudinary;
import com.cloudinary.utils.ObjectUtils;
import com.mini.socialnetwork.model.Comment;
import com.mini.socialnetwork.model.Post;
import com.mini.socialnetwork.repository.CommentRepository;
import com.mini.socialnetwork.repository.PostRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class CommentService {

    private static final long MAX_IMAGE_SIZE_BYTES = 5 * 1024 * 1024; // 5MB

    private final CommentRepository commentRepository;
    private final PostRepository postRepository;
    private final Cloudinary cloudinary;

    /**
     * Create a new comment on a post
     */
    public Comment createComment(String postId, String userId, String content, MultipartFile image) throws IOException {
        UUID postUuid = UUID.fromString(postId);
        UUID userUuid = UUID.fromString(userId);

        // Validate post exists and is not deleted
        Post post = postRepository.findById(postUuid)
                .orElseThrow(() -> new IllegalArgumentException("Post not found"));
        if (post.isDeleted()) {
            throw new IllegalArgumentException("Cannot comment on a deleted post");
        }

        // Validate content or image exists
        boolean hasContent = StringUtils.hasText(content);
        boolean hasImage = image != null && !image.isEmpty();

        if (!hasContent && !hasImage) {
            throw new IllegalArgumentException("Comment must have content or an image");
        }

        // Upload image to Cloudinary if present
        String imageUrl = null;
        if (hasImage) {
            if (image.getSize() > MAX_IMAGE_SIZE_BYTES) {
                throw new IllegalArgumentException("Image must be 5MB or smaller");
            }
            Map<?, ?> uploadResult = cloudinary.uploader()
                    .upload(image.getBytes(), ObjectUtils.asMap("folder", "comments"));
            Object url = uploadResult.get("secure_url");
            if (url != null) {
                imageUrl = url.toString();
            }
        }

        // Build and save comment
        Comment comment = Comment.builder()
                .postId(postUuid)
                .userId(userUuid)
                .content(hasContent ? content : null)
                .imageUrl(imageUrl)
                .createdAt(Instant.now())
                .updatedAt(Instant.now())
                .isDeleted(false)
                .build();

        Comment saved = commentRepository.save(comment);

        // Update post comment count
        long commentCount = commentRepository.countByPostIdAndIsDeletedFalse(postUuid);
        post.setCommentCount((int) commentCount);
        post.setUpdatedAt(Instant.now());
        postRepository.save(post);

        return saved;
    }

    /**
     * Get a comment by ID
     */
    public Comment getCommentById(String commentId) {
        UUID uuid = UUID.fromString(commentId);
        Comment comment = commentRepository.findById(uuid)
                .orElseThrow(() -> new IllegalArgumentException("Comment not found"));
        if (comment.isDeleted()) {
            throw new IllegalArgumentException("Comment not found");
        }
        return comment;
    }

    /**
     * Get comments for a post with pagination
     */
    public Slice<Comment> getCommentsByPost(String postId, int page, int size) {
        UUID postUuid = UUID.fromString(postId);
        Pageable pageable = PageRequest.of(page, size);
        return commentRepository.findByPostIdAndIsDeletedFalseOrderByCreatedAtDesc(postUuid, pageable);
    }

    /**
     * Update a comment (only owner can update)
     */
    public Comment updateComment(String commentId, String userId, String content, MultipartFile image,
            boolean removeImage) throws IOException {
        UUID commentUuid = UUID.fromString(commentId);
        UUID userUuid = UUID.fromString(userId);

        Comment comment = commentRepository.findById(commentUuid)
                .orElseThrow(() -> new IllegalArgumentException("Comment not found"));

        if (comment.isDeleted()) {
            throw new IllegalArgumentException("Comment not found");
        }

        // Check ownership
        if (!comment.getUserId().equals(userUuid)) {
            throw new IllegalArgumentException("You can only edit your own comments");
        }

        // Update content if provided
        boolean hasNewContent = StringUtils.hasText(content);
        if (hasNewContent) {
            comment.setContent(content);
        }

        // Handle image update
        boolean hasNewImage = image != null && !image.isEmpty();
        if (hasNewImage) {
            if (image.getSize() > MAX_IMAGE_SIZE_BYTES) {
                throw new IllegalArgumentException("Image must be 5MB or smaller");
            }
            Map<?, ?> uploadResult = cloudinary.uploader()
                    .upload(image.getBytes(), ObjectUtils.asMap("folder", "comments"));
            Object url = uploadResult.get("secure_url");
            if (url != null) {
                comment.setImageUrl(url.toString());
            }
        } else if (removeImage) {
            comment.setImageUrl(null);
        }

        // Validate final state has content or image
        if (!StringUtils.hasText(comment.getContent()) && comment.getImageUrl() == null) {
            throw new IllegalArgumentException("Comment must have content or an image");
        }

        comment.setUpdatedAt(Instant.now());
        return commentRepository.save(comment);
    }

    /**
     * Soft delete a comment (only owner can delete)
     */
    public Comment deleteComment(String commentId, String userId) {
        UUID commentUuid = UUID.fromString(commentId);
        UUID userUuid = UUID.fromString(userId);

        Comment comment = commentRepository.findById(commentUuid)
                .orElseThrow(() -> new IllegalArgumentException("Comment not found"));

        if (comment.isDeleted()) {
            throw new IllegalArgumentException("Comment not found");
        }

        // Check ownership
        if (!comment.getUserId().equals(userUuid)) {
            throw new IllegalArgumentException("You can only delete your own comments");
        }

        comment.setDeleted(true);
        comment.setUpdatedAt(Instant.now());
        Comment deleted = commentRepository.save(comment);

        // Update post comment count
        UUID postUuid = comment.getPostId();
        Post post = postRepository.findById(postUuid).orElse(null);
        if (post != null) {
            long commentCount = commentRepository.countByPostIdAndIsDeletedFalse(postUuid);
            post.setCommentCount((int) commentCount);
            post.setUpdatedAt(Instant.now());
            postRepository.save(post);
        }

        return deleted;
    }

    /**
     * Get comment count for a post
     */
    public long getCommentCount(String postId) {
        UUID postUuid = UUID.fromString(postId);
        return commentRepository.countByPostIdAndIsDeletedFalse(postUuid);
    }
}

File: MiniSocialNetwork_Fe/src/app/core/services/comment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

/* ===== DTO (Data Transfer Objects) ===== */
export interface CommentResponse {
    id: string;
    postId: string;
    userId: string;
    content: string | null;
    imageUrl: string | null;
    createdAt: string;
    updatedAt: string;
    deleted: boolean;
}

export interface SliceResponse<T> {
    content: T[];
    hasNext: boolean;
}

/* ===== COMMENT SERVICE ===== */
@Injectable({
    providedIn: 'root',
})
export class CommentService {
    private readonly API_URL = `${environment.apiUrl}/comments`;

    constructor(private http: HttpClient) { }

    /**
     * Tạo comment mới (hỗ trợ upload ảnh)
     * POST /api/comments
     * Form data: postId, userId, content (optional), image (optional)
     */
    createComment(
        postId: string,
        userId: string,
        content: string | null,
        image: File | null
    ): Observable<CommentResponse> {
        const formData = new FormData();
        formData.append('postId', postId);
        formData.append('userId', userId);

        if (content && content.trim().length > 0) {
            formData.append('content', content.trim());
        }

        if (image) {
            formData.append('image', image);
        }

        return this.http.post<CommentResponse>(this.API_URL, formData);
    }

    /**
     * Lấy comment theo ID
     * GET /api/comments/{id}
     */
    getCommentById(commentId: string): Observable<CommentResponse> {
        return this.http.get<CommentResponse>(`${this.API_URL}/${commentId}`);
    }

    /**
     * Lấy danh sách comments của một post với phân trang (Slice)
     * GET /api/comments/post/{postId}?page=0&size=10
     */
    getCommentsByPost(
        postId: string,
        page = 0,
        size = 10
    ): Observable<SliceResponse<CommentResponse>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString());

        return this.http.get<SliceResponse<CommentResponse>>(
            `${this.API_URL}/post/${postId}`,
            { params }
        );
    }

    /**
     * Cập nhật comment (chỉ owner mới có quyền)
     * PUT /api/comments/{id}
     * Form data: userId, content (optional), image (optional), removeImage (optional)
     */
    updateComment(
        commentId: string,
        userId: string,
        content: string | null,
        image: File | null,
        removeImage = false
    ): Observable<CommentResponse> {
        const formData = new FormData();
        formData.append('userId', userId);

        if (content && content.trim().length > 0) {
            formData.append('content', content.trim());
        }

        if (image) {
            formData.append('image', image);
        }

        if (removeImage) {
            formData.append('removeImage', 'true');
        }

        return this.http.put<CommentResponse>(
            `${this.API_URL}/${commentId}`,
            formData
        );
    }

    /**
     * Xóa comment (soft delete, chỉ owner mới có quyền)
     * DELETE /api/comments/{id}?userId=xxx
     */
    deleteComment(
        commentId: string,
        userId: string
    ): Observable<CommentResponse> {
        const params = new HttpParams().set('userId', userId);
        return this.http.delete<CommentResponse>(`${this.API_URL}/${commentId}`, {
            params,
        });
    }

    /**
     * Lấy số lượng comment của một post
     * GET /api/comments/count/{postId}
     */
    getCommentCount(postId: string): Observable<number> {
        return this.http.get<number>(`${this.API_URL}/count/${postId}`);
    }
}
